frame/system/src/lib.rs

```diff
@@ -953,6 +953,28 @@ impl<
 	}
 }
 
+/// The "AND gate" implementation of `EnsureOrigin`.
+///
+/// Origin check will pass if `L` and `R` origin check passes. `L` is tested first.
+pub struct EnsureBothOf<AccountId, L, R>(sp_std::marker::PhantomData<(AccountId, L, R)>);
```

Comment:
I've actually been thinking about whether it's a good thing to introduce such a new struct, or simply implement `EnsureOrigin` on tuples using impl_for_tuple. That way we're not constrained on just 2 origins, but can specify as many `EnsureOrigin`s as we want.

Reply:
I am not sure that it will be better than such implementation of `EnsureBothOf` or `EnsureOneOf`, because it will not so flexible as with the usage if these ones.

Reply:
I'm sorry, but why would it be less flexible? Compare `EnsureBothOf<AccountId, EnsureRoot, EnsureBothOf<AccountId, EnsureSigned, EnsureFounder>>` and `(EnsureRoot, EnsureSigned, EnsureFounder)`, you're telling me that the former is more flexible than the latter? It does not look so to me.

Reply:
At this case yes, but I am looking into possibility to combine `EnsureBothOf` with `EnsureOneOf` and leave an under stable interface.
But I like your idea how to make such struct with the different amount arguments for both of them `EnsureBothOf`, `EnsureOneOf`. Will try to implement it.

Reply:
Using tuples to imply `Or` is 100% more readable than nested types.

Reply:
note that we can have EnsureOneOf with as maybe generic as we want. no need for nested stuff.
We can create a type with default generic
```rust
pub struct EnsureOneOf<A, B, C=ImpossibleOrigin, D=ImpossibleOrigin, ...>
```
I feel like EnsureOneOf is more clear than a tuple personally